### DISCUSSION TITLE
Use different programming language for UAD-ng

### DISCUSSION THREAD

**Discussion by code-kungfu** [Universal-Debloater-Alliance/universal-android-debloater-next-generation#349] [Ideas]
Rust has a steep learning curve and its build system isn't friendly, not to mention the lack of solid UI libraries.
It does not add any benefits.

Porting UAD to Modern Object Pascal have many benefits:
* Easy to learn, yet powerful and modern programming language
* Solid integration to platform native widget sets (Tap into Win32, GTK, QT, Cocoa etc)
* Truly a write once, compile anywhere toolchain - which is blazing fast.
* The IDE (Lazarus that is) makes it easy to quickly design and tweak the UI
* Compiler targets for virtually anything
* Huge community
* Huge set of libraries and components
* (Yes, you can write memory safe code too)

The TL;DR here is you get a faster development cycle and make the project more approachable for contributions.

**Comment by Frigyes06** [comment]
I strongly oppose this proposal. You are making claims that are unreasonable and untrue. How would switching to Pascal give the project a faster development cycle? We'll have to release more patches because someone wrote unsafe code (again)?
Also, you are talking about a DEBLOATING project. Offering an IDE, the definition of bloated to this community is not a wise thing to do.
You also claim the build system isn't friendly. I guess typing cargo build is too complicated for some folks.
The only fair point you have is the questionable UI support, and the question of should we even be writing UIs in rust? Either way, we are, it works, and it's decent. https://areweguiyet.com/
Using rust has tremendeous benefits both for our contributors and our end users too. I do not see this project ever switching to another language, especially not Pascal.

PS:
You clearly have some personal bias against rust that you'll have to work out, as your entire internet persona is based in hating on rust. "Master of the great Modern Object Pascal language. Compiler guy, next generation Object Pascal lang. Wielder of electrolysis to expel Rust from your system." Really?

**Reply by code-kungfu** [reply]
You're clearly uninformed @Frigyes06  about Modern Object Pascal.
You can write safe code too in Modern Object Pascal - Then go explain why it is on the recommended list of programming languages for writing safe code released by NSA?
The language itself and the compiler holds you to strict coding standards which enforces good programming practices and reduce what you would deem "unsafe code".

Lazarus / Free Pascal is the IDE and compiler to quickly write and compile such a project - it's far quicker to install and get up and running than Rust.

Modern Object Pascal has a much lower learning curve than Rust - this is a fact. Contributors will have a much easier time contributing to the utility, and, you have the opportunity to design a more user friendly and streamlined UI than it is the case now.
Given the compiler itself is much faster than Rust, the write - compile - debug cycle is faster.

Cargo is a pain to deal with for many - so yes, it is indeed a problem.

I would suggest you go use Modern Object Pascal first before you dig yourself deeper.

PS: I'm glad you're recognizing my tongue in cheek comment on my profile, but, seeing you're brining this into the conversation it is just an Ad Hominem argument 🙃

**Reply by Frigyes06** [reply]
Feb 2024 White House release: https://www.whitehouse.gov/wp-content/uploads/2024/02/Final-ONCD-Technical-Report.pdf

Also, why would you ever take advice from the NSA without a grain of salt? The last thing they want is to have software so secure they can't infiltrate it, if needed :)

This is my last comment in this discussion.

**Reply by code-kungfu** [reply]
I'm aware of the WH report as well, it is just a recommendation.
As for the NSA report here: https://media.defense.gov/2023/Apr/27/2003210083/-1/-1/0/CSI_SOFTWARE_MEMORY_SAFETY_V1.1.PDF
It is a recommendation as well.

That said, Modern Object Pascal sill delivers on the premise of writing safer code, however the whole "memory safe" factor is an illusion and honestly, Rust wont save us, but it ideas will.
Rust just adds unnecessary complexity to the stack.

**Comment by AnonymousWP** [comment]
Edited title to make it more generic.

**Comment by vitoramaral10** [comment]
I may be talking garbage and not know much about UAD itself, but wouldn't Flutter be a viable option? 
A few points I'm making here:
- it's multi-platform
- easy to program
- Very simple and customizable UI (Material)
- Active and growing community

Now I don't know if you need something specific to run UAD, if you can let me know I can research it.

**Reply by AnonymousWP** [reply]
Thanks. At the moment, we're not prioritising switching to another language, but if we will, we'll take your comment into consideration.

**Comment by Rudxain** [comment]
https://www.hylo-lang.org

Of course, we should only consider using Hylo in the far future. It's too young
